build/Shared/TaskResultCache.cs

```diff
@@ -0,0 +1,113 @@
+        /// </summary>
+        /// <param name="key">The key for the async operation to get or store in the cache.</param>
+        /// <param name="refresh"><see langword="true" /> to force the specified asynchronous operation to be executed and stored in the cache even if a cached operation exists, otherwise <see langword="false" />.</param>
+        /// <param name="valueFactory">A <see cref="Func{TResult}" /> to execute asynchronously if a cached operation does not exist.</param>
+        /// <returns>A <see cref="Task{TResult}" /> for the specified asynchronous operation from the cache if found, otherwise the scheduled asynchronous operation to await.</returns>
+        public Task<TValue> GetOrAddAsync(TKey key, bool refresh, Func<Task<TValue>> valueFactory)
+        {
+            return GetOrAddAsync(key, refresh, valueFactory, CancellationToken.None);
+        }
+
+        /// <summary>
+        /// Gets the cached async operation associated with the specified key, or runs the operation asynchronously and returns <see cref="Task{TValue}" /> that the caller can await, and optionally refreshes the cache.
+        /// </summary>
+        /// <param name="key">The key for the async operation to get or store in the cache.</param>
+        /// <param name="refresh"><see langword="true" /> to force the specified asynchronous operation to be executed and stored in the cache even if a cached operation exists, otherwise <see langword="false" />.</param>
+        /// <param name="valueFactory">A <see cref="Func{TResult}" /> to execute asynchronously if a cached operation does not exist.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken" /> to use for signaling that an operation should be cancelled.</param>
+        /// <returns>A <see cref="Task{TResult}" /> for the specified asynchronous operation from the cache if found, otherwise the scheduled asynchronous operation to await.</returns>
+        public async Task<TValue> GetOrAddAsync(TKey key, bool refresh, Func<Task<TValue>> valueFactory, CancellationToken cancellationToken)
+        {
+            AsyncLazy<TValue> lazy = new(() => valueFactory()
+                    .ContinueWith(
+                        (task, state) => task.GetAwaiter().GetResult(),
+                        state: null,
+                        cancellationToken,
+                        TaskContinuationOptions.RunContinuationsAsynchronously,
+                        TaskScheduler.Default));
+
+            AsyncLazy<TValue> result = refresh
+                ? _cache.AddOrUpdate(key, lazy, (_, _) => lazy)
+                : _cache.GetOrAdd(key, lazy);
+
+            return await result;
+        }
+
+#pragma warning disable VSTHRD011 // Lazy<Task<T>>.Value can deadlock. Use AsyncLazy<T> instead, which we are
+        public class AsyncLazy<T> : Lazy<Task<T>>
```

Comment:
At the moment I have `TaskResultCache` in `build\Shared` so its compiled into everything and it can be `internal`.  Not all of our assemblies reference `NuGet.Common` so those won't compile if they reference that `AsyncLazy<T>` of course.  I could put `TaskResultCache` into `NuGet.Common` but then it would have to be public.  Thoughts?

Reply:
Let's leave it as is. 
Let's just have a targetted fix are better, as long as you're aware of the other implementation.